File: go/200number-of-islands/200.岛屿数量.go
/*
 * @lc app=leetcode.cn id=200 lang=golang
 *
 * [200] 岛屿数量
 */

// @lc code=start
func numIslands(grid [][]byte) int {
	result := 0
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[i]); j++ {
			if grid[i][j] == '1' {
				result++;
				destroyIsland(grid, i, j)
			}
		}
	}
	return result
}

func destroyIsland(grid [][]byte, i int, j int) {
	if (i < 0 || i > len(grid) - 1 || j < 0 || j > len(grid[i]) - 1) {
		return
	}
	if (grid[i][j] == '1') {
		grid[i][j] = '0'
	} else {
		return
	}
	
	destroyIsland(grid, i - 1, j)
	destroyIsland(grid, i + 1, j)
	destroyIsland(grid, i, j - 1)
	destroyIsland(grid, i, j + 1)
}
// @lc code=end

File: java/200numIslands/200.岛屿数量.java
/*
 * @lc app=leetcode.cn id=200 lang=java
 *
 * [200] 岛屿数量
 */

// @lc code=start
class Solution {
    public int numIslands(char[][] grid) {
        if(grid == null || grid.length == 0) {
            return 0;
        }
        int islandCount = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if(grid[i][j] == '1') {
                    islandCount++;
                    destroyIsland(grid, i, j);
                }
            }
        }
        return islandCount;
    }

    private void destroyIsland(char[][] grid, int i, int j) {
        if(grid[i][j] == '0') {
            return;
        }
        grid[i][j] = '0';
        if(i+1 < grid.length) {
            destroyIsland(grid, i+1, j);
        }
        if(i-1 >= 0) {
            destroyIsland(grid, i-1, j);
        }
        if(j+1 < grid[i].length) {
            destroyIsland(grid, i, j+1);
        }
        if(j-1 >= 0) {
            destroyIsland(grid, i, j-1);
        }
    }
}
// @lc code=end

